enterprise/server/remote_execution/snaploader/snaploader.go

```diff
@@ -207,6 +234,89 @@ func (l *FileCacheLoader) CacheSnapshot(ctx context.Context, key *fcpb.SnapshotK
 	return &Snapshot{key: key, manifest: manifest}, nil
 }
 
+func (l *FileCacheLoader) unpackCOW(ctx context.Context, file *fcpb.ChunkedFile, outputDirectory string) (cow *blockio.COW, err error) {
+	dataDir := filepath.Join(outputDirectory, file.GetName())
+	if err := os.Mkdir(dataDir, 0755); err != nil {
+		return nil, status.InternalErrorf("failed to create COW data dir %q: %s", dataDir, err)
+	}
+	var chunks []*blockio.Chunk
+	defer func() {
+		// If there was an error, clean up any chunks we created.
+		if err == nil {
+			return
+		}
+		for _, c := range chunks {
+			c.Close()
+		}
+	}()
+	for _, chunk := range file.Chunks {
+		size := file.GetChunkSize()
+		if remainder := file.GetSize() - chunk.GetOffset(); size > remainder {
+			size = remainder
+		}
+		d := &repb.Digest{Hash: chunk.GetDigestHash(), SizeBytes: size}
+		node := &repb.FileNode{Digest: d}
+		path := filepath.Join(dataDir, fmt.Sprintf("%d", chunk.GetOffset()))
+		if !l.env.GetFileCache().FastLinkFile(node, path) {
+			return nil, status.UnavailableErrorf("snapshot chunk %s/%d not found in local cache", file.GetName(), chunk.GetOffset())
+		}
+		mm, err := blockio.NewLazyMmap(path)
+		if err != nil {
+			return nil, status.WrapError(err, "create mmap for chunk")
+		}
+		c := &blockio.Chunk{Offset: chunk.GetOffset(), Store: mm}
+		chunks = append(chunks, c)
+	}
+	return blockio.NewCOW(chunks, file.GetChunkSize(), file.GetSize(), dataDir)
+}
+
+func (l *FileCacheLoader) cacheCOW(ctx context.Context, name string, cow *blockio.COW) (*fcpb.ChunkedFile, error) {
+	size, err := cow.SizeBytes()
+	if err != nil {
+		return nil, err
+	}
+	cf := &fcpb.ChunkedFile{
+		Name:      name,
+		Size:      size,
+		ChunkSize: cow.ChunkSizeBytes(),
+	}
+	for i, c := range cow.Chunks {
+		if _, ok := c.(*blockio.Hole); ok {
+			continue
+		}
+		if cow.Dirty(i) {
+			// Sync dirty chunks to make sure the underlying file is up to date
+			// before we add it to cache.
+			if err := c.Sync(); err != nil {
+				return nil, status.WrapError(err, "sync dirty chunk")
+			}
+		}
+		// TODO: It's too expensive to compute sha256 digests here for every
+		// chunk. Try out the following improvements:
```

Comment:
Would be great to have a benchmark here :P so we know what too expensive means...